### PR TITLE
New: [AEA-4497] - deploy cognito

### PR DESCRIPTION
**Pull request by anthony-nhs**
## Summary

- Routine Change

### Details

- deploy resources for cognito

### CODE REVIEW

File: .vscode/eps-prescription-tracker-ui.code-workspace [hunk 1]
```diff
@@ -19,11 +19,31 @@
...
+        "path": "../packages/cognito"
+      },
+      {
+        "name": "packages/common/middyErrorHandler",
+        "path": "../packages/common/middyErrorHandler"
+      },
+      {
+        "name": "packages/common/testing",
+        "path": "../packages/common/testing"
+      }
```
Comment (kris-szlapa) [inline .vscode/eps-prescription-tracker-ui.code-workspace:34]: Missing comma after curly bracket.

File: .github/workflows/cdk_package_code.yml [hunk 2]
```diff
@@ -60,7 +61,7 @@ jobs:
             package-lock.json \
             tsconfig.defaults.json \
             Makefile \
-            cdk.json
+            cdk.json
```
Comment (kris-szlapa) [inline .github/workflows/cdk_package_code.yml:64]: Unnecessary whitespace at the end of line 64.

Comment (anthony-nhs) [reply .github/workflows/cdk_package_code.yml:64]: fixed

**Review by kris-szlapa — DISMISSED** [review]

**Review by kris-szlapa — APPROVED** [review]

File: packages/auth_demo/src/App.test.tsx [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+import { render, screen } from '@testing-library/react';
+import App from './App';
+
+test('renders learn react link', () => {
```
Comment (MatthewPopat-NHS) [inline packages/auth_demo/src/App.test.tsx:5]: does it?

File: packages/auth_demo/src/App.tsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      const accessToken = authSession.tokens?.accessToken
+      const idToken = authSession.tokens?.idToken
+      if (accessToken && idToken) {
+        console.log(idToken.payload)
+        setAccessToken(accessToken)
+        setIdToken(idToken)
+        setIsSignedIn(true)
+        const currentUser = await getCurrentUser();
+        setUser(currentUser);
+        }
```
Comment (MatthewPopat-NHS) [inline packages/auth_demo/src/App.tsx:52]: Indentation gone awry here

File: packages/auth_demo/src/App.tsx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      console.log(payload)
+      switch (payload.event) {
+        case "signInWithRedirect":
+          getUser();
+          break;
+        case "signInWithRedirect_failure":
+          setError("An error has ocurred during the OAuth flow.");
+          break;
+        case "customOAuthState":
+          setCustomState(payload.data); // this is the customState provided on signInWithRedirect function
```
Comment (MatthewPopat-NHS) [inline packages/auth_demo/src/App.tsx:30]: odd that we're capturing these and not showing them anywhere

File: packages/auth_demo/src/index.tsx [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+root.render(
+  <React.StrictMode>
+    <App />
+  </React.StrictMode>
+);
+
+// If you want to start measuring performance in your app, pass a function
+// to log results (for example: reportWebVitals(console.log))
+// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
+reportWebVitals();
```
Comment (MatthewPopat-NHS) [inline packages/auth_demo/src/index.tsx:19]: This seems like an unnecessary dependency to be dragging along

File: packages/cdk/resources/Cognito.ts [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    if (props.useMockOidc) {
+      if (props.mockOidcAuthorizeEndpoint === undefined ||
+        props.mockOidcjwksEndpoint === undefined ||
+        props.mockOidcUserInfoEndpoint === undefined ||
+        props.mockOidcClientId === undefined ||
+        props.mockOidClientSecret === undefined ||
+        props.mockOidcIssuer === undefined
+      ) {
+        throw new Error("Attempt to use mock oidc but variables are not defined")
+      }
```
Comment (MatthewPopat-NHS) [inline packages/cdk/resources/Cognito.ts:127]: This seems a bit OTT, we're not doing the same checks for the primary props.

File: packages/cdk/resources/Cognito.ts [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+          OAuthScope.OPENID,
+          OAuthScope.EMAIL,
+          OAuthScope.PHONE,
+          OAuthScope.PROFILE,
+          OAuthScope.COGNITO_ADMIN
+        ],
+        callbackUrls: [
+          "http://localhost:3000/auth/",
+          `https://${props.fullCloudfrontDomain}/site/`,
+          `https://${props.fullCloudfrontDomain}/auth_demo/`
```
Comment (MatthewPopat-NHS) [inline packages/cdk/resources/Cognito.ts:185]: maybe a TODO comment to remove these when we remove auth demo

File: packages/cdk/resources/Cognito.ts [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+      clientSecret: props.primaryOidClientSecret,
+      issuerUrl: props.primaryOidcIssuer,
+      userPool: userPool,
+      attributeRequestMethod: OidcAttributeRequestMethod.GET,
+      scopes: ["openid", "profile", "email", "nhsperson", "nationalrbacaccess"],
+      endpoints: oidcEndpoints
+    })
+
+    const supportedIdentityProviders: Array<UserPoolClientIdentityProvider> = [
+      UserPoolClientIdentityProvider.COGNITO,
```
Comment (MatthewPopat-NHS) [inline packages/cdk/resources/Cognito.ts:110]: Why are we allowing Cognito itself to be an identity provider? I thought we said that real CIS2 and mock auth would be all we need.

File: packages/cdk/resources/Cognito.ts [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+          `https://${props.fullCloudfrontDomain}/site/`,
+          `https://${props.fullCloudfrontDomain}/auth_demo/`
+        ]
+      }})
+
+    // ensure dependencies are set correctly so items are created in the correct order
+    userPoolWebClient.node.addDependency(primaryPoolIdentityProvider)
+    if (props.useMockOidc) {
+      userPoolWebClient.node.addDependency(mockPoolIdentityProvider)
+    }
```
Comment (MatthewPopat-NHS) [inline packages/cdk/resources/Cognito.ts:198]: This seems like overkill, they are directly referenced in the definition of the client, do we definitely need the explicit dependency as well?

File: packages/cdk/resources/CognitoFunctions.ts [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    // set up things for mock authorization
+    if (props.useMockOidc) {
+      if (props.mockOidcjwksEndpoint === undefined ||
+        props.mockOidcUserInfoEndpoint === undefined ||
+        props.mockOidcTokenEndpoint === undefined ||
+        props.mockOidcClientId === undefined ||
+        props.mockOidcIssuer === undefined
+      ) {
+        throw new Error("Attempt to use mock oidc but variables are not defined")
+      }
```
Comment (MatthewPopat-NHS) [inline packages/cdk/resources/CognitoFunctions.ts:118]: as above

File: packages/cdk/resources/CognitoFunctions.ts [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+            resources: ["*"]
+          }),
+          new PolicyStatement({
+            effect: Effect.ALLOW,
+            principals: [props.deploymentRole],
+            actions: [
+              "kms:Encrypt",
+              "kms:GenerateDataKey*"
+            ],
+            resources:["*"]
```
Comment (MatthewPopat-NHS) [inline packages/cdk/resources/CognitoFunctions.ts:86]: as above

File: packages/cdk/resources/CognitoFunctions.ts [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+          })
+        ]
+      })
+    })
+    jwtKmsKey.addAlias(`alias/${props.stackName}-jwtKmsKey`)
```
Comment (MatthewPopat-NHS) [inline packages/cdk/resources/CognitoFunctions.ts:91]: this is creating a duplicate alias, the "alias" property directly on the key is already making one.

File: packages/cdk/resources/CognitoFunctions.ts [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+      secretStringValue: SecretValue.unsafePlainText("ChangeMe"),
+      encryptionKey: jwtKmsKey
+    })
+    primaryJwtPrivateKey.addToResourcePolicy( new PolicyStatement({
+      effect: Effect.ALLOW,
+      principals: [props.deploymentRole],
+      actions: [
+        "secretsmanager:PutSecretValue"
+      ],
+      resources:["*"]
```
Comment (MatthewPopat-NHS) [inline packages/cdk/resources/CognitoFunctions.ts:180]: as above

File: packages/cdk/resources/Dynamodb/dynamodbResources.ts [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+ * Common resources for a dynamodb table
+
+ */
+
+export interface DynamodbResourcesProps {
+  readonly stackName: string
+  readonly table: TableV2
+}
+
+export class DynamodbResources extends Construct {
```
Comment (MatthewPopat-NHS) [inline packages/cdk/resources/Dynamodb/dynamodbResources.ts:15]: I'm not sure why we're separating these into their own file. Why not just declare them in packages/cdk/resources/Dynamodb.ts

File: packages/cdk/resources/RestApiGateway.ts [hunk 16]
```diff
@@ -81,12 +76,11 @@ export class RestApiGateway extends Construct {
...
-    /* Dummy Method/Resource to allow a gateway to be deployed "empty" */
-    apiGateway.root.addMethod("ANY", new MockIntegration({
-      integrationResponses: [
-        {statusCode: "418"}
-      ]
-    }))
+    const authorizer = new CognitoUserPoolsAuthorizer(this, "Authorizer", {
+      authorizerName: "cognitoAuth",
+      cognitoUserPools: [props.userPool],
+      identitySource: "method.request.header.authorization"
```
Comment (MatthewPopat-NHS) [inline packages/cdk/resources/RestApiGateway.ts:82]: I think this is the default, we could be explicit but I think I would prefer specify the minimum we can

File: packages/cdk/resources/Cognito.ts (hunk 9 above)
Comment (anthony-nhs) [reply packages/cdk/resources/Cognito.ts:110]: removed

File: packages/cdk/resources/Cognito.ts (hunk 7 above)
Comment (anthony-nhs) [reply packages/cdk/resources/Cognito.ts:127]: this is needed for typing. 
The props in the if statement are optional to this construct and are only passed in when useMockOidc is true. 
If we dont have the if statement, then typescript thinks they may be undefined and so throws errors when compiling the resources inside the if statement

File: packages/cdk/resources/Cognito.ts (hunk 10 above)
Comment (anthony-nhs) [reply packages/cdk/resources/Cognito.ts:198]: This is one of the times where explicit dependencies are needed to ensure resources are created in the correct order

File: packages/cdk/resources/CognitoFunctions.ts [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+          new PolicyStatement({
+            sid: "Enable IAM User Permissions",
+            effect: Effect.ALLOW,
+            actions: [
+              "kms:*"
+            ],
+            principals: [
+              new AccountRootPrincipal
+            ],
+            resources: ["*"]
```
[the remark this answers is not in the document]
Comment (anthony-nhs) [reply packages/cdk/resources/CognitoFunctions.ts:77]: it is not meaningless - it is required. See resources section of https://docs.aws.amazon.com/kms/latest/developerguide/key-policy-overview.html#key-policy-elements

File: packages/cdk/resources/CognitoFunctions.ts (hunk 12 above)
Comment (anthony-nhs) [reply packages/cdk/resources/CognitoFunctions.ts:86]: it is not meaningless - it is required. See resources section of https://docs.aws.amazon.com/kms/latest/developerguide/key-policy-overview.html#key-policy-elements

File: packages/cdk/resources/CognitoFunctions.ts (hunk 13 above)
Comment (anthony-nhs) [reply packages/cdk/resources/CognitoFunctions.ts:91]: removed

File: packages/cdk/resources/CognitoFunctions.ts (hunk 11 above)
Comment (anthony-nhs) [reply packages/cdk/resources/CognitoFunctions.ts:118]: this is needed for typing.
The props in the if statement are optional to this construct and are only passed in when useMockOidc is true.
If we dont have the if statement, then typescript thinks they may be undefined and so throws errors when compiling the resources inside the if statement

File: packages/cdk/resources/CognitoFunctions.ts (hunk 14 above)
Comment (anthony-nhs) [reply packages/cdk/resources/CognitoFunctions.ts:180]: it is not meaningless - it is required. See resources section of https://docs.aws.amazon.com/kms/latest/developerguide/key-policy-overview.html#key-policy-elements

File: packages/cdk/resources/RestApiGateway.ts (hunk 16 above)
Comment (anthony-nhs) [reply packages/cdk/resources/RestApiGateway.ts:82]: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_apigateway.CognitoUserPoolsAuthorizer.html#identitysource reads as though it is required

File: packages/cdk/resources/Dynamodb/dynamodbResources.ts (hunk 15 above)
Comment (anthony-nhs) [reply packages/cdk/resources/Dynamodb/dynamodbResources.ts:15]: fixed

File: packages/cdk/resources/Cognito.ts [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+        },
+        scopes: [
+          OAuthScope.OPENID,
+          OAuthScope.EMAIL,
+          OAuthScope.PHONE,
+          OAuthScope.PROFILE,
+          OAuthScope.COGNITO_ADMIN
+        ],
+        callbackUrls: [
+          "http://localhost:3000/auth/",
```
[the remark this answers is not in the document]
Comment (anthony-nhs) [reply packages/cdk/resources/Cognito.ts:182]: fixed

File: packages/cdk/resources/RestApiGateway.ts (hunk 16 above)
Comment (MatthewPopat-NHS) [reply packages/cdk/resources/RestApiGateway.ts:82]: identitySource?
Type: string (optional, default: IdentitySource.header('Authorization'))

File: packages/auth_demo/src/App.tsx (hunk 4 above)
Comment (MatthewPopat-NHS) [reply packages/auth_demo/src/App.tsx:52]: decided to leave auth_demo stuff as it is temporary.

File: packages/auth_demo/src/index.tsx (hunk 6 above)
Comment (MatthewPopat-NHS) [reply packages/auth_demo/src/index.tsx:19]: decided to leave auth_demo stuff as it is temporary.
